client/signup/config/steps-pure.js

```diff
@@ -166,7 +166,9 @@ export function generateSteps( {
 			stepName: 'plans',
 			apiRequestFunction: addPlanToCart,
 			dependencies: [ 'siteSlug' ],
-			providesDependencies: [ 'cartItem' ],
+			optionalIncomingDependencies: [ 'siteId', 'domainItem', 'newSiteParams' ],
```

Comment:
This is a little confusing, because if it's optional, then by definition the step does not really have it as a "dependency" :smile: 

I see why you did this, though -- you needed to pass this stuff along to the API request function, and the dependencies are really the only way to do that without a larger refactoring.

Two thoughts:
1. I would say at a minimum, if we're going to add this, let's document it in `client/signup/README.md` along with the other properties.
2. Also, an alternate suggestion, could something like this work instead?:

```
dependencies: [ 'siteSlug', 'siteId', 'domainItem', 'newSiteParams' ],
dependencyDefaults: {
  siteId: null,
  domainItem: {},
  newSiteParams: {},
},
```

The idea is that everything would still go in `dependencies` (and the step could always assume that all dependencies are present), but the code that builds the list of dependencies would use `dependencyDefaults` to fill in any missing properties before passing them along.

To me this would be easier to understand and probably more flexible for the future too.

Reply:
I'm not sold on the dependencyDefaults idea because the flow controller relies on the mere existing of the keys in the dependencies object to consider them "fulfilled", even if their value is null or undefined, I think it may also lead to some confusion because we would be basically values that do not exist in the signup progress store.

I'm more inclined to maintain `optionalIncomingDependencies` (name suggestions are welcome) or just pass the whole object of dependencies to the processing function, I don't see how the extra properties would even affect them. I'd like to know @niranjan-uma-shankar thoughts about this, since he's more familiar with the signup flows.

Reply:
Just an fyi that we already use `defaultDependencies`in a few places. 

https://github.com/Automattic/wp-calypso/blob/master/client/signup/config/steps-pure.js#L202